pyop2/backends/cuda.py

```diff
@@ -0,0 +1,625 @@
+            return self._data
+
+    @property
+    def data_ro(self):
+        if len(self._data) == 0:
+            raise RuntimeError("Illegal access: No data associated with"
+                               " this Global!")
+
+        if cuda_backend.offloading:
+            self.ensure_availability_on_device()
+            view = self._cuda_data.view()
+            # FIXME: PyCUDA doesn't support read-only arrays yet
+            return view
+        else:
+            self.ensure_availability_on_host()
+            view = self._data.view()
+            view.setflags(write=False)
+            return view
+
+
+@dataclass(frozen=True)
+class CUFunctionWithExtraArgs:
+    cu_func: cuda.Function
+    args_to_append: Tuple[cuda.DeviceAllocation, ...]
+
+    def __call__(self, grid, block, *args):
+        if all(grid + block):
+            self.cu_func.prepared_call(grid, block,
+                                       *args,
+                                       *[arg_to_append.gpudata
+                                         for arg_to_append in self.args_to_append])
+
+
+class GlobalKernel(AbstractGlobalKernel):
+    @classmethod
+    def _cache_key(cls, *args, **kwargs):
+        key = super()._cache_key(*args, **kwargs)
+        key = key + (configuration["gpu_strategy"],)
+        return key
+
+    @utils.cached_property
+    def encoded_cache_key(self):
+        return md5(str(self.cache_key[1:]).encode()).hexdigest()
+
+    @utils.cached_property
+    def computational_grid_expr_cache_file_path(self):
+        cachedir = configuration["cache_dir"]
+        return os.path.join(cachedir, f"{self.encoded_cache_key}_grid_params")
+
+    @utils.cached_property
+    def extra_args_cache_file_path(self):
+        cachedir = configuration["cache_dir"]
+        return os.path.join(cachedir, f"{self.encoded_cache_key}_extra_args.npz")
+
+    @memoize_method
+    def get_grid_size(self, start, end):
+        from pymbolic import parse, evaluate
+
+        fpath = self.computational_grid_expr_cache_file_path
+
+        with open(fpath, "r") as f:
+            glens, llens = f.read().split("\n")
+
+        glens = parse(glens)
+        llens = parse(llens)
+
+        parameters = {"start": start, "end": end}
+
+        grid = tuple(int(evaluate(glens[i], parameters)) if i < len(glens) else 1
+                     for i in range(2))
+        block = tuple(int(evaluate(llens[i], parameters)) if i < len(llens) else 1
+                      for i in range(3))
+
+        return grid, block
+
+    @memoize_method
+    def get_extra_args(self):
+        """
+        Returns device buffers corresponding to array literals baked into
+        :attr:`local_kernel`.
+        """
+        from pytools import natsorted
+        fpath = self.extra_args_cache_file_path
+        extra_args_np = [arg
+                         for _, arg in natsorted(numpy.load(fpath).items(),
+                                                 key=lambda x: x[0])]
```

Comment:
This is _tremendously_ fragile. Please think of a better way of saving this data in a way that doesn't rely on dubiously-stable properties of numpy.savez/load.

Reply:
Agreed. While `np.savez`-ing the arrays also stored a string-array that tells us the keys that were used for storing these arrays. See https://github.com/OP2/PyOP2/compare/9b7eb97878f2b6609d64affe896c5d4d901a2934..0e58f08d3ea1487e01c45df2270fa9e5790a845d.